src/chainstate/stacks/miner.rs

```diff
@@ -56,6 +56,149 @@ use util::vrf::*;
 use core::*;
 use core::mempool::*;
 
+///
+///    Independent structure for building microblocks:
+///       StacksBlockBuilder cannot be used, since microblocks should only be broadcasted
+///       once the anchored block is mined, won sortition, and a StacksBlockBuilder will
+///       not survive that long.
+///
+///     StacksMicroblockBuilder holds a mutable reference to the provided chainstate in the
+///       new function. This is required for the `clarity_tx` -- basically, to append transactions
+///       as new microblocks, the builder _needs_ to be able to keep the current clarity_tx "open" 
+pub struct StacksMicroblockBuilder<'a> {
+    anchor_block: BlockHeaderHash,
+    anchor_block_bhh: BurnchainHeaderHash,
+    anchor_block_height: u64,
+    prev_microblock_header: Option<StacksMicroblockHeader>,
+    header_reader: StacksChainState,
+    clarity_tx: Option<ClarityTx<'a>>,
+    considered: Option<HashSet<Txid>>,
+}
+
+impl <'a> StacksMicroblockBuilder <'a> {
+    pub fn new(anchor_block: BlockHeaderHash, anchor_block_bhh: BurnchainHeaderHash,
+               chainstate: &'a mut StacksChainState, initial_cost: ExecutionCost) -> Result<StacksMicroblockBuilder<'a>, Error> {
+        let header_reader = chainstate.reopen()?;
+        let mut clarity_tx = chainstate.block_begin(&anchor_block_bhh, &anchor_block,
+                                                    &MINER_BLOCK_BURN_HEADER_HASH, &MINER_BLOCK_HEADER_HASH);
+        let anchor_block_height = 
+            StacksChainState::get_anchored_block_header_info(&header_reader.headers_db, &anchor_block_bhh, &anchor_block)?
+            .ok_or(Error::NoSuchBlockError)?
+            .block_height;
+
+        clarity_tx.reset_cost(initial_cost);
+        Ok(StacksMicroblockBuilder {
+            anchor_block,
+            anchor_block_bhh,
+            anchor_block_height,
+            clarity_tx: Some(clarity_tx),
+            header_reader,
+            prev_microblock_header: None,
+            considered: Some(HashSet::new()),
+        })
+    }
+
+    pub fn mine_next_microblock(&mut self,
+                                mem_pool: &MemPoolDB,
+                                miner_key: &Secp256k1PrivateKey,
+                                miner_pubkey_hash: &Hash160) -> Result<StacksMicroblock, Error> {
```

Comment:
Should there be a maximum microblock size?  Or do we only care about the size of the entire epoch's data?  I'm leaning more towards the latter, but could be convinced of the former.  Either way, we'll need to track the total epoch size.

Reply:
Yes -- I think it needs to be included in total epoch size.

Reply:
Updated in 808abde